### PR TITLE
[RNET-992] Add sync support for collections in mixed

### PR DESCRIPTION
**Pull request by rorbech**
## Description

This adds sync support for collections in mixed fields.

Fixes [RNET-992](https://jira.mongodb.org/browse/RNET-992)

##  TODO

* [ ] ~~Changelog entry~~ Collections in mixed is already part of the changelog.
* [x] Tests

### CODE REVIEW

File: Tools/DeployApps/BaasClient.cs [hunk 1]
```diff
@@ -148,6 +148,7 @@ public class FunctionReturn
             };";
 
         private readonly HttpClient _client = new();
+        private readonly HttpClient _privateclient = new();
```
Comment (papafe) [inline Tools/DeployApps/BaasClient.cs:151]: We can remove all of this now, right?

File: .github/workflows/pr.yml [hunk 2]
```diff
@@ -39,6 +39,44 @@ jobs:
     - build-wrappers
     with:
       build-docs: ${{ contains(github.head_ref, 'release') }}
+  publish-packages-to-sleet:
```
Comment (nirinchev) [inline .github/workflows/pr.yml:42]: Is this temporary?

File: Tools/DeployApps/BaasClient.cs [hunk 3]
```diff
@@ -556,6 +561,14 @@ private async Task<(BaasApp App, string MongoServiceId)> CreateAppCore(string na
...
             });
 
+            // Retrieve feature flags
+            // var features = await GetPrivateAsync<BsonDocument>($"groups/{_groupId}/apps/{app}/features");
+            // ENABLE LEGACY MIXED SUPPORT
+            // await PostPrivateAsync<BsonDocument>($"features/legacy_mixed_type", new { app_ids = new []{$"{appId}"}, action = "enable"} );
+            // await PostPrivateAsync<BsonDocument>($"features/bypass_legacy_mixed_type", new { app_ids = new []{$"{appId}"}, action = "disable"} );
+            // DISABLE LEGACY MIXED SUPPORT - Default on latest test server - https://github.com/10gen/baas/blob/master/etc/configs/test_config.json#L303
+            // await PostPrivateAsync<BsonDocument>($"features/legacy_mixed_type", new { app_ids = new []{$"{appId}"}, action = "disable"} );
+            // await PostPrivateAsync<BsonDocument>($"features/bypass_legacy_mixed_type", new { app_ids = new []{$"{appId}"}, action = "enable"} );
```
Comment (nirinchev) [inline Tools/DeployApps/BaasClient.cs:571]: We probably don't need this. While it's likely useful information, by the time we need it again, no one will remember it's here.

File: .github/workflows/pr.yml (hunk 2 above)
Comment (papafe) [reply .github/workflows/pr.yml:42]: Didn't we say we wanted to publish the PR packages to sleet?

Comment (nirinchev) [reply .github/workflows/pr.yml:42]: Oh sorry, I must have not been clear here - I meant we wanted to do a one-off publish so that the server team can add a dependency on it for their e2e tests on CI. We don't need to do it on a continuous basis and definitely not for all PRs.

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs [hunk 4]
```diff
@@ -514,8 +586,333 @@ private void TestPropertyCore<T>(Func<SyncAllTypesObject, T> getter, Action<Sync
...
+                await realm2.SyncSession.WaitForUploadAsync();
+                await realm1.SyncSession.WaitForDownloadAsync();
+                await realm2.SyncSession.WaitForDownloadAsync();
+
+                var list1 = obj1.RealmValueProperty.AsDictionary()["list"].AsList();
+                var dictionary1 = obj1.RealmValueProperty.AsDictionary()["dictionary"].AsDictionary();
+                var list2 = obj1.RealmValueProperty.AsDictionary()["list"].AsList();
+                var dictionary2 = obj1.RealmValueProperty.AsDictionary()["dictionary"].AsDictionary();
+
+                CollectionAssert.DoesNotContain(list1, (RealmValue)1);
```
Comment (papafe) [inline Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs:905]: Can't you use the `Assert.That` style with those?

**Review by papafe — COMMENTED** [review]
Overall it looks good! Just some comments

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs [hunk 5]
```diff
@@ -275,21 +285,73 @@ public class DataTypeSynchronizationTests : SyncTestBase
...
+                    "string",
+                    new byte[] { 0, 1, 2 },
+                    new DateTimeOffset(1234, 5, 6, 7, 8, 9, TimeSpan.Zero),
+                    1f,
+                    2d,
+                    3m,
+                    new ObjectId("5f63e882536de46d71877979"),
+                    Guid.Parse("3809d6d9-7618-4b3d-8044-2aa35fd02f31"),
+                    // new InternalObject { IntProperty = 10, StringProperty = "brown" },
+                    // innerList,
```
[the remark this answers is not in the document]
Comment (rorbech) [reply Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs]: Just forgot to include them after polishing the tests.

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs [hunk 6]
```diff
@@ -514,8 +586,333 @@ private void TestPropertyCore<T>(Func<SyncAllTypesObject, T> getter, Action<Sync
...
+                {
+                    realm1.Subscriptions.Add(realm1.All<SyncAllTypesObject>());
+                    realm1.Subscriptions.Add(realm1.All<IntPropertyObject>());
+                });
+
+                var child = new IntPropertyObject();
+                var valuesValue = new List<RealmValue>()
+                {
+                    1,
+                    (RealmValue)"Realm",
```
[the remark this answers is not in the document]
Comment (rorbech) [reply Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs]: Just followed the pattern from elsewhere, but seems that there is now implicit conversion in place 😎

File: CHANGELOG.md [hunk 7]
```diff
@@ -96,7 +96,7 @@
 * Realm Studio: 15.0.0 or later.
 
 ### Internal
-* Using Core 14.1.0.
+* Using Core 14.4.1.
```
Comment (papafe) [inline CHANGELOG.md:99]: Forgot this before, but we should also update the changelog with relevant core changes.

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs [hunk 8]
```diff
@@ -236,7 +246,7 @@ public class DataTypeSynchronizationTests : SyncTestBase
         public void Dict_Binary() => TestDictionaryCore(o => o.ByteArrayDict, TestHelpers.GetBytes(10), TestHelpers.GetBytes(15), (a, b) => a.SequenceEqual(b));
 
         [Test]
-        public void Property_Binary() => TestPropertyCore(o => o.ByteArrayProperty, (o, rv) => o.ByteArrayProperty = rv, TestHelpers.GetBytes(5), TestHelpers.GetBytes(10), (a, b) => a!.SequenceEqual(b!));
+        public void Property_Binary() => TestPropertyCore(o => o.ByteArrayProperty, (o, rv) => o.ByteArrayProperty = rv, TestHelpers.GetBytes(5), TestHelpers.GetBytes(10));
```
Comment (papafe) [inline Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs:249]: Why was this changed?

File: CHANGELOG.md (hunk 7 above)
Comment (rorbech) [reply CHANGELOG.md:99]: Ok. Think it makes sense for @papafe  to do it, since you can better judge what to include.

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs (hunk 8 above)
Comment (rorbech) [reply Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs:249]: I adopted the pattern of using `Assert.That` with `Is.EqualTo().Using()` with the `RealmValueWithCollections.RealmValueComparer`. I guess it wraps non-RealmValues in RealmValues, but considered it easier than abstracting the comparison just for this single case.

File: CHANGELOG.md (hunk 7 above)
Comment (papafe) [reply CHANGELOG.md:99]: Nikola just updated the changelog and core in another PR, so no need to do it

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs [hunk 9]
```diff
@@ -22,16 +22,26 @@
...
+using Nito.AsyncEx;
 using NUnit.Framework;
+using NUnit.Framework.Constraints;
+using Realms.Exceptions;
 using Realms.Extensions;
 using Realms.Helpers;
+using Realms.Logging;
+using Realms.Sync;
+using Realms.Sync.Exceptions;
+using Realms.Tests.Database;
```
Comment (papafe) [inline Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs:36]: I don't think all of these are used, are they?

**Review by papafe — APPROVED** [review]
Looks good to me! Just left a small comment

File: Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs (hunk 9 above)
Comment (rorbech) [reply Tests/Realm.Tests/Sync/DataTypeSynchronizationTests.cs:30]: The comparer is from `Realms.Tests.Database` but there were a lot of unused `using`. I have cleaned it up.
